test/find-golden-tickets.py

```diff
@@ -69,51 +73,70 @@ def post_bundle(bundle_fqid, es_query, indexer_url):
 
 
 def main(argv: List[str]):
-    """
-    Entrypoint method for the script
-    """
     args = parser.parse_args(argv)
     dss_client = DSSClient()
     dss_client.host = args.dss_url
     # noinspection PyUnresolvedReferences
     response = dss_client.post_search.iterate(es_query=args.es_query, replica="aws")
     bundle_fqids = [r['bundle_fqid'] for r in response]
+    logger.info("Bundle FQIDs to index: %i", len(bundle_fqids))
+
     errors = defaultdict(int)
     missing = {}
     indexed = 0
     total = 0
-    for bundle_fqid in bundle_fqids:
-        total += 1
-        print(f"Bundle: {bundle_fqid}")
-        retries = 3
-        while True:
+    futures = []
+
+    with ThreadPoolExecutor(max_workers=args.num_workers, thread_name_prefix='pool') as tpe:
+
+        def attempt(bundle_fqid, i):
             try:
+                logger.info("Bundle %s, attempt %i: Sending notification", bundle_fqid, i)
                 post_bundle(bundle_fqid=bundle_fqid,
                             es_query=args.es_query,
                             indexer_url=args.indexer_url)
-                indexed += 1
-            except HTTPError as er:
-                # Current retry didn't work. Try again
-                print(f"Error sending bundle to indexer:\n{er}")
-                print(f"{retries} retries left")
-                if retries > 0:
-                    retries -= 1
-                    sleep(retries)
+            except HTTPError as e:
+                if i < 3:
+                    logger.warning("Bundle %s, attempt %i: scheduling retry after error %s", bundle_fqid, i, e)
+                    return bundle_fqid, tpe.submit(partial(attempt, bundle_fqid, i + 1))
                 else:
-                    print("Out of retries, there might be a problem.")
-                    print(f"Error:\n{er}")
-                    errors[er.code] += 1
-                    missing[bundle_fqid] = er.code
-                    break
+                    logger.warning("Bundle %s, attempt %i: giving up after error %s", bundle_fqid, i, e)
+                    return bundle_fqid, e
             else:
-                break
-    print(f"Total of bundle_fqids read: {total}")
-    print(f"Total of {indexed} bundle_fqids indexed")
-    print("Total number of errors by code:")
-    pprint(dict(errors))
-    print("Missing bundle_fqids and their error code:")
-    pprint(missing)
+                logger.info("Bundle %s, attempt %i: success", bundle_fqid, i)
+                return bundle_fqid, None
+
+        def handle_future(future):
+            nonlocal indexed
+            exception = future.exception()
```

Comment:
@hannes-ucsc make comment explaining it blocks

Reply:
fixed